Make document.applets always return an empty list

annevk opened the pull request:
#### 2ca55a3e19df60127dec09f3af22e4d3ab2943ec
<pre>
Make document.applets always return an empty list
<a href="https://bugs.webkit.org/show_bug.cgi?id=255326">https://bugs.webkit.org/show_bug.cgi?id=255326</a>
rdar://107926196

Reviewed by Myles C. Maxfield.

Replace DocApplets with DocEmpty that never matches any node. Also remove containsJavaApplet() as it&apos;s no longer called by anything.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/leaktests/html-collection-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/leaktests/html-collection.html:

Upstreamed via <a href="https://github.com/web-platform-tests/wpt/pull/39499.">https://github.com/web-platform-tests/wpt/pull/39499.</a>

* Source/WebCore/dom/Document.cpp:
(WebCore::canAccessAncestor):
(WebCore::Document::sharedLogger):
(WebCore::Document::removedLastRef):
(WebCore::Document::adoptNode):
(WebCore::Document::setTitle):
(WebCore::Document::resolveStyle):
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
(WebCore::Document::removeAllEventListeners):
(WebCore::Document::axObjectCache const):
(WebCore::Document::hasHighlight const):
(WebCore::Document::updateHighlightPositions):
(WebCore::Document::implicitClose):
(WebCore::Document::setURL):
(WebCore::Document::processReferrerPolicy):
(WebCore::Document::runScrollSteps):
(WebCore::Document::updateIsPlayingMedia):
(WebCore::Document::focusNavigationStartingNode const):
(WebCore::Document::applets):
(WebCore::Document::finishedParsing):
(WebCore::Document::statePopped):
(WebCore::Document::initDNSPrefetch):
(WebCore::Document::resumeScheduledTasks):
(WebCore::Document::absoluteEventHandlerBounds):
(WebCore::Document::updateLastHandledUserGestureTimestamp):
(WebCore::Document::loader const):
(WebCore::Document::convertAbsoluteToClientRects):
(WebCore::Document::convertAbsoluteToClientRect):
(WebCore::expandRootBoundsWithRootMargin):
(WebCore::computeClippedRectInRootContentsSpace):
(WebCore::Document::contentChangeObserver):
* Source/WebCore/html/CollectionType.h:
* Source/WebCore/html/GenericCachedHTMLCollection.cpp:
(WebCore::GenericCachedHTMLCollection&lt;traversalType&gt;::elementMatches const):
* Source/WebCore/html/HTMLCollection.cpp:
(WebCore::HTMLCollection::rootTypeFromCollectionType):
(WebCore::invalidationTypeExcludingIdAndNameAttributes):
(WebCore::HTMLCollection::isSupportedPropertyName):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::containsJavaApplet const): Deleted.
* Source/WebCore/html/HTMLObjectElement.h:

Canonical link: <a href="https://commits.webkit.org/262879@main">https://commits.webkit.org/262879@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84b2ead56d53225f9f4f6a780099b670d6073622

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2881 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4295 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3030 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2990 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/2537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3297 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2577 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4085 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2559 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2430 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2553 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2610 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3834 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2959 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2604 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2562 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/714 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2593 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2780 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8685,7 +8685,7 @@ Logger& Document::logger()
 
     return *m_logger;
 }
-    
```
Were all these really necessary to do in this patch?

cdumez replied:
This adds so much noise to include this in patches...

annevk committed:
Make document.applets always return an empty list
https://bugs.webkit.org/show_bug.cgi?id=255326
rdar://107926196

Reviewed by Myles C. Maxfield.

Replace DocApplets with DocEmpty that never matches any node. Also remove containsJavaApplet() as it's no longer called by anything.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/leaktests/html-collection-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/leaktests/html-collection.html:

Upstreamed via https://github.com/web-platform-tests/wpt/pull/39499.

* Source/WebCore/dom/Document.cpp:
(WebCore::canAccessAncestor):
(WebCore::Document::sharedLogger):
(WebCore::Document::removedLastRef):
(WebCore::Document::adoptNode):
(WebCore::Document::setTitle):
(WebCore::Document::resolveStyle):
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
(WebCore::Document::removeAllEventListeners):
(WebCore::Document::axObjectCache const):
(WebCore::Document::hasHighlight const):
(WebCore::Document::updateHighlightPositions):
(WebCore::Document::implicitClose):
(WebCore::Document::setURL):
(WebCore::Document::processReferrerPolicy):
(WebCore::Document::runScrollSteps):
(WebCore::Document::updateIsPlayingMedia):
(WebCore::Document::focusNavigationStartingNode const):
(WebCore::Document::applets):
(WebCore::Document::finishedParsing):
(WebCore::Document::statePopped):
(WebCore::Document::initDNSPrefetch):
(WebCore::Document::resumeScheduledTasks):
(WebCore::Document::absoluteEventHandlerBounds):
(WebCore::Document::updateLastHandledUserGestureTimestamp):
(WebCore::Document::loader const):
(WebCore::Document::convertAbsoluteToClientRects):
(WebCore::Document::convertAbsoluteToClientRect):
(WebCore::expandRootBoundsWithRootMargin):
(WebCore::computeClippedRectInRootContentsSpace):
(WebCore::Document::contentChangeObserver):
* Source/WebCore/html/CollectionType.h:
* Source/WebCore/html/GenericCachedHTMLCollection.cpp:
(WebCore::GenericCachedHTMLCollection<traversalType>::elementMatches const):
* Source/WebCore/html/HTMLCollection.cpp:
(WebCore::HTMLCollection::rootTypeFromCollectionType):
(WebCore::invalidationTypeExcludingIdAndNameAttributes):
(WebCore::HTMLCollection::isSupportedPropertyName):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::containsJavaApplet const): Deleted.
* Source/WebCore/html/HTMLObjectElement.h:

Canonical link: https://commits.webkit.org/262879@main

cdumez commented on Source/WebCore/html/GenericCachedHTMLCollection.cpp:
```diff
@@ -74,10 +74,10 @@ bool GenericCachedHTMLCollection<traversalType>::elementMatches(Element& element
-        return is<HTMLObjectElement>(element) && downcast<HTMLObjectElement>(element).containsJavaApplet();
     case DocEmbeds:
         return element.hasTagName(embedTag);
+    case DocEmpty:
```
This seems very inefficient. As far as I can tell, this will traverse the whole DOM tree and return false for every element. We should not be traversing the DOM tree at all or even have any caching logic.

rniwa replied:
Yeah, we probably want to introduce a new traversal type that's basically a no-op.

cdumez replied:
If we introduced a new very simple `EmptyHTMLCollection` (similar to `EmptyNodeList`) which subclasses `HTMLCollection` and overrides:
```
Element* item(unsigned) const final { return nullptr; }
Element* namedItem(const AtomString& name) const final { return nullptr; }
unsigned length() const final { return 0; }
```

Then no traversal at all.

rniwa replied:
Yeah, but we don't even need to subclass it. We can just create a new CollectionTraversal. Let me post a PR in a bit.

rniwa replied:
Fixing this in https://bugs.webkit.org/show_bug.cgi?id=255389.

cdumez reviewed (commented):
This doesn't seem great. Odds are we want a new EmptyHTMLCollection subclass for HTMLCollection (similarly to the pre-existing EmptyNodeList).
